src/fluree/db/migrations/nameservice_v1.cljc

```diff
@@ -172,16 +179,188 @@
+                             ledger-path (str seg1 "/" seg2)]
+                         {:old-path p
+                          :alias    (str ledger-path ":" branch)
+                          :new-path (str const/ns-version "/" ledger-path "/" branch ".json")})
+
+                       (re-matches v1-nested-one-regex p)
+                       (let [[_ ledger branch] (re-matches v1-nested-one-regex p)]
+                         {:old-path p
+                          :alias    (str ledger ":" branch)
+                          :new-path (str const/ns-version "/" ledger "/" branch ".json")})
+
+                       :else nil)))
+             vec))
+      [])))
+
+(defn migrate-v1-file
+  "Migrate one ns@v1 nameservice file to ns@v2 nested layout. Idempotent if new file exists."
+  [file-store {:keys [old-path new-path alias]}]
+  (go-try
+    (log/info "Migrating ns@v1 record for" alias ":" old-path "->" new-path)
+    (if (<? (storage/read-bytes file-store new-path))
+      (do
+        (log/info "New path already exists for" alias ", removing old file only")
+        (let [addr (-> file-store storage/location (storage/build-address old-path))]
+          (<? (storage/delete file-store addr)))
+        {:migrated false :skipped true :alias alias :old-path old-path :new-path new-path})
+      (let [content (<? (storage/read-bytes file-store old-path))
+            bytes*  (cond
+                      (nil? content) nil
+                      (string? content) (bytes/string->UTF8 content)
+                      :else content)]
+        (when-not bytes*
+          (throw (ex-info (str "Unable to read nameservice file: " old-path)
+                          {:status 500 :error :db/migration})))
+        (<? (storage/write-bytes file-store new-path bytes*))
+        (let [addr (-> file-store storage/location (storage/build-address old-path))]
+          (<? (storage/delete file-store addr)))
+        (log/info "Migrated ns@v1 record for" alias)
+        {:migrated true :alias alias :old-path old-path :new-path new-path}))))
+
+(defn migrate-v1-to-v2
+  "Migrate all ns@v1 files (flat and nested) to ns@v2 nested layout."
+  [file-store]
+  (go-try
+    (let [v1-files (<? (find-v1-nameservice-files file-store))]
+      (if (seq v1-files)
+        (do
+          (log/info "Found" (count v1-files) "ns@v1 files to migrate to" const/ns-version)
+          (loop [files v1-files
+                 results []]
+            (if-let [info (first files)]
+              (let [res (<? (migrate-v1-file file-store info))]
+                (recur (rest files) (conj results res)))
+              (do
+                (log/info "ns@v1 ->" const/ns-version "migration completed. Migrated"
+                          (count (filter :migrated results))
+                          "files, skipped" (count (filter :skipped results)) "already updated files")
+                results))))
+        []))))
+
 (defn run-migration-if-needed
```

Comment:
If a fluree db is configured to use a non-file store backend (s3 or ipfs), will _not_ running this migration have the system be in a usable state? Or are those configurations not supported until we add migrations for them?

Reply:
IPFS isn't working in this version, and S3 was just added and I'm quite certain I'm the only one using it.